src/Microsoft.TemplateEngine.Edge/Settings/GlobalSettings.cs

```diff
@@ -1,42 +1,198 @@
+
+        public event Action? SettingsChanged;
+
+        private List<TemplatesSourceData> _userInstalledTemplatesSources = new List<TemplatesSourceData>();
 
-        [JsonIgnore]
         public IReadOnlyList<TemplatesSourceData> UserInstalledTemplatesSources => _userInstalledTemplatesSources;
 
-        // What this does, if older TemplateEngine loads this file and save it back
-        // it will include new settings that new TemplateEngine depends on
-        // without this field, data would be lost in process of loading and saving
-        //[JsonExtensionData]
-        //private IDictionary<string, JToken> _additionalData;
         public void Add(TemplatesSourceData userInstalledTemplate)
         {
+            if (!_locked)
+                throw new InvalidOperationException($"Call {nameof(LockAsync)} before calling this method");
             _userInstalledTemplatesSources.RemoveAll(data => data.MountPointUri == userInstalledTemplate.MountPointUri);
             _userInstalledTemplatesSources.Add(userInstalledTemplate);
             SettingsChanged?.Invoke();
         }
 
         public void Remove(TemplatesSourceData userInstalledTemplate)
         {
+            if (!_locked)
+                throw new InvalidOperationException($"Call {nameof(LockAsync)} before calling this method");
             _userInstalledTemplatesSources.RemoveAll(data => data.MountPointUri == userInstalledTemplate.MountPointUri);
             SettingsChanged?.Invoke();
         }
+
+        class DisposableCallback : IDisposable
+        {
+            private Action<Stream, Mutex>? _disposeCalled;
+            private Stream _fileStream;
+            private Mutex _mutex;
+
+            public DisposableCallback(Action<Stream, Mutex> disposeCalled, Stream fileStream, Mutex mutex)
+            {
+                _disposeCalled = disposeCalled;
+                _fileStream = fileStream;
+                _mutex = mutex;
+            }
+
+            public void Dispose()
+            {
+                var disposeCalled = Interlocked.Exchange(ref _disposeCalled, null);
+                disposeCalled?.Invoke(_fileStream, _mutex);
+            }
+        }
+
+        public async Task<IDisposable> LockAsync(CancellationToken token)
+        {
+            while (true)
+            {
+                if (token.IsCancellationRequested)
+                    throw new TaskCanceledException();
+
+                try
+                {
+                    var stream = _environmentSettings.Host.FileSystem.CreateFileStream(_globalSettingsFile, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.Read);
+                    var mutex = new Mutex(false, "{DEAEF4B4-4B59-4A80-A488-9D85CC18818A}");
+                    if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    {
+                        while (!token.IsCancellationRequested)
+                        {
+                            if (await Task.Run(() => mutex.WaitOne(20)))
```

Comment:
consider this overload of Run to avoid hangs:
https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.run?view=net-5.0#System_Threading_Tasks_Task_Run_System_Action_System_Threading_CancellationToken_

Reply:
Changed, this logic, I'm aware of this overload, but for some reason not using it as often :(